lib/block.ml

```diff
@@ -365,28 +368,42 @@ let mk ~line ~file ~column ~section ~labels ~legacy_labels ~header ~contents
       match header with
       | Some Header.OCaml ->
           let file_kind = Fk_ocaml { part_included = part } in
-          Ok (Include { file_included; file_kind })
+          Ok (`Include { file_included; file_kind })
       | _ ->
           check_not_set "`part` is not supported for non-OCaml code blocks."
             part
           >>= fun () ->
           let file_kind = Fk_other { header } in
-          Ok (Include { file_included; file_kind }) )
+          Ok (`Include { file_included; file_kind }) )
   | None -> (
       check_not_set "`part` label requires a `file` label." part >>= fun () ->
       match header with
       | Some Header.Shell ->
           check_no_errors errors >>= fun () ->
           check_not_set "`env` label cannot be used with a `shell` header." env
-          >>= fun () -> Ok (Cram { non_det })
+          >>= fun () -> Ok (`Cram { non_det })
       | Some Header.OCaml -> (
           let env = Env.mk env in
           match guess_ocaml_kind contents with
-          | `Code -> Ok (OCaml { env; non_det; errors })
+          | `Code -> Ok (`OCaml { env; non_det; errors })
           | `Toplevel ->
               check_no_errors errors >>= fun () ->
-              Ok (Toplevel { env; non_det }) )
-      | _ -> check_no_errors errors >>= fun () -> Ok (Raw { header }) ) )
+              Ok (`Toplevel { env; non_det }) )
+      | _ -> check_no_errors errors >>= fun () -> Ok (`Raw { header }) ) )
+  >>= fun guess ->
+  ( match (guess, block_kind) with
+  | `Include b, (None | Some Include) -> Ok (Include b)
```

Comment:
I'm not sure that's how it should work. I think the explicit block kind should prevail no matter what.

If no block kind is specified, then we try to infer what it is based on the header, the labels, the syntax etc.

One of the main reason why those are useful is that they allow for clear error reporting and easy to document behaviour.

Reply:
I wrote it this way because we have to check the block labels whether the block kind is specified or not. Let's say I set the block kind to `toplevel` I have to check all the labels corresponding to the record fields are set (to fill the record), and that none of the incompatible labels are set. That's why I check the labels first, and then check that what is guessed from them is consistent with the block kind if it exists.

Reply:
I haven't taken a deep enough look but that seems dubious to me. I feel like the `guess` is already the result of the inference. That means that the error will always be the one from inference. Also you if the inferred block differs from the explicit one, the error is always the same. I feel like that's not really the intended behaviour.

Could you add tests to make sure this behaves properly?

Reply:
I redesigned the whole code, I still have to add some tests.